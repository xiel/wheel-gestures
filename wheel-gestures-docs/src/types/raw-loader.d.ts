declare module '!!raw-loader!*' {
  const content: string
  export = content
}
