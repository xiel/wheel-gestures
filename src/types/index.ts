/**
 * public types
 * these types are exported by the module
 */
export type VectorXYZ = [number, number, number]
export type BooleanXYZ = [boolean, boolean, boolean]
export type PreventWheelActionType = true | 'x' | 'y' | 'z' | false
export type ReverseSign = BooleanXYZ | boolean

export interface WheelGesturesConfig {
  preventWheelAction: PreventWheelActionType
  reverseSign: ReverseSign
}

// all options are optional and have reasonable defaults
export type WheelGesturesOptions = Partial<WheelGesturesConfig>
export type WheelEventDataRequiredFields = 'deltaMode' | 'deltaX' | 'deltaY' | 'timeStamp'

export interface WheelEventData
  extends Pick<WheelEvent, WheelEventDataRequiredFields>,
    Partial<Omit<WheelEvent, WheelEventDataRequiredFields>> {}

export interface WheelEventState {
  isStart: boolean
  isMomentum: boolean
  isEnding: boolean
  isMomentumCancel: boolean
  axisMovement: VectorXYZ
  axisVelocity: VectorXYZ
  axisDelta: VectorXYZ
  event: WheelEvent | WheelEventData
  previous?: WheelEventState
}

export type WheelGesturesEventMap = {
  wheel: WheelEventState
}
