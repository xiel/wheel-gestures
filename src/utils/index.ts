export * from './projection'

export function lastOf<T>(array: T[]) {
  return array[array.length - 1]
}

export function average(numbers: number[]) {
  return numbers.reduce((a, b) => a + b) / numbers.length
}

export const clamp = (value: number, min: number, max: number) => Math.min(Math.max(min, value), max)

export function addVectors<T extends number[]>(v1: T, v2: T): T {
  if (v1.length !== v2.length) {
    throw new Error('vectors must be same length')
  }
  return v1.map((val, i) => val + v2[i]) as T
}

export function absMax(numbers: number[]) {
  return Math.max(...numbers.map(Math.abs))
}

// eslint-disable-next-line @typescript-eslint/ban-types
export function deepFreeze<T extends object>(o: T): Readonly<T> {
  Object.freeze(o)
  Object.values(o).forEach((value) => {
    if (value !== null && typeof value === 'object' && !Object.isFrozen(value)) {
      deepFreeze(value)
    }
  })
  return o
}
