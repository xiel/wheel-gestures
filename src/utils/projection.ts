const DECAY = 0.996

/**
 * movement projection based on velocity
 * @param velocityPxMs
 * @param decay
 */
export const projection = (velocityPxMs: number, decay = DECAY) => (velocityPxMs * decay) / (1 - decay)
