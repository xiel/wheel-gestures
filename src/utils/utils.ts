export function lastOf<T>(array: T[]) {
  return array[array.length - 1]
}

export function firstOf<T>(array: T[]) {
  return array[0]
}
