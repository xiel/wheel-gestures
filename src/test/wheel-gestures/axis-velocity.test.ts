import { WheelEventState } from '../../types'
import { lastOf } from '../../utils'
import { generateEvents } from '../helper/generateEvents'
import { subscribeAndFeedWheelEvents } from '../helper/recordPhases'

function calcAverageVelocity(phases: WheelEventState[]) {
  return phases.reduce(
    (averageAcc, { axisVelocity }) => {
      averageAcc[0] = (averageAcc[0] + axisVelocity[0]) / 2
      averageAcc[1] = (averageAcc[1] + axisVelocity[1]) / 2
      return averageAcc
    },
    [0, 0]
  )
}

describe('velocity', () => {
  it('should have velocity in first (and only) event - x', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents({
      wheelEvents: [{ deltaX: 0, deltaY: 20, deltaMode: 0, timeStamp: 0 }],
    })
    const [xVelo, yVelo] = allPhaseData[1].axisVelocity

    // velocity in first event is based on the initial timeout between two events
    expect(xVelo).toEqual(0)
    expect(yVelo).toEqual(20 / 400)
    expect(allPhaseData).toMatchSnapshot()
  })

  it('check average velocity', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents(generateEvents({ deltaTotal: [0, 500], durationMs: 1000 }))
    const averageVelocity = calcAverageVelocity(allPhaseData)

    expect(averageVelocity[0]).toBeCloseTo(0)
    expect(averageVelocity[1]).toBeCloseTo(0.5)
  })

  it('check average velocity - different total delta', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents(generateEvents({ deltaTotal: [0, 1000], durationMs: 1000 }))
    const averageVelocity = calcAverageVelocity(allPhaseData)

    expect(averageVelocity[0]).toBeCloseTo(0)
    expect(averageVelocity[1]).toBeCloseTo(1)
  })

  it('velocity & axisDeltas - different event rate', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents(
      generateEvents({ deltaTotal: [0, 1000], durationMs: 1000, eventEveryMs: 1000 / 30 })
    )
    const averageVelocity = calcAverageVelocity(allPhaseData)

    expect(averageVelocity[0]).toBeCloseTo(0)
    expect(averageVelocity[1]).toBeCloseTo(1)

    const [deltaX, deltaY] = lastOf(allPhaseData).axisMovement

    expect(deltaX).toBeCloseTo(0)
    expect(deltaY).toBeCloseTo(1000)
  })

  it('velocity & axisDeltas - x & y axis', () => {
    const { allPhaseData } = subscribeAndFeedWheelEvents(
      generateEvents({ deltaTotal: [2000, 2000], durationMs: 1000, eventEveryMs: 1000 / 30 })
    )
    const averageVelocity = calcAverageVelocity(allPhaseData)

    expect(averageVelocity[0]).toBeCloseTo(2)
    expect(averageVelocity[1]).toBeCloseTo(2)

    const [deltaX, deltaY] = lastOf(allPhaseData).axisMovement

    expect(deltaX).toBeCloseTo(2000)
    expect(deltaY).toBeCloseTo(2000)
  })
})
