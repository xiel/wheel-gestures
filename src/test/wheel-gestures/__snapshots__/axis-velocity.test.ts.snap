// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`velocity should have velocity in first (and only) event - x 1`] = `
Array [
  Object {
    "axisDelta": Array [
      0,
      20,
      0,
    ],
    "axisMovement": Array [
      0,
      20,
      0,
    ],
    "axisMovementProjection": Array [
      0,
      32.44999999999999,
      0,
    ],
    "axisVelocity": Array [
      0,
      0.05,
      0,
    ],
    "event": Object {
      "deltaMode": 0,
      "deltaX": 0,
      "deltaY": 20,
      "timeStamp": 0,
    },
    "isEnding": false,
    "isMomentum": false,
    "isMomentumCancel": false,
    "isStart": true,
    "previous": undefined,
  },
  Object {
    "axisDelta": Array [
      0,
      0,
      0,
    ],
    "axisMovement": Array [
      0,
      20,
      0,
    ],
    "axisMovementProjection": Array [
      0,
      32.44999999999999,
      0,
    ],
    "axisVelocity": Array [
      0,
      0.05,
      0,
    ],
    "event": Object {
      "deltaMode": 0,
      "deltaX": 0,
      "deltaY": 20,
      "timeStamp": 0,
    },
    "isEnding": true,
    "isMomentum": false,
    "isMomentumCancel": false,
    "isStart": false,
    "previous": Object {
      "axisDelta": Array [
        0,
        20,
        0,
      ],
      "axisMovement": Array [
        0,
        20,
        0,
      ],
      "axisMovementProjection": Array [
        0,
        32.44999999999999,
        0,
      ],
      "axisVelocity": Array [
        0,
        0.05,
        0,
      ],
      "event": Object {
        "deltaMode": 0,
        "deltaX": 0,
        "deltaY": 20,
        "timeStamp": 0,
      },
      "isEnding": false,
      "isMomentum": false,
      "isMomentumCancel": false,
      "isStart": true,
    },
  },
]
`;
